### PR TITLE
Topic sk jcs 13724 & jcs 13758

### PR DESCRIPTION
**Pull request by skommala**
MRs for jiras :
JCS-13724 : ER 33432789 Run network validation script when provisioning WLS for OCI with existing subnets & JCS-13758
 : Modifications needed to the network validation scripts before implementing ER 33432789

Tests run
---------
Existing subnets, existing nsgs , and bastion
Existing subnets, existing nsgs, and lb with regional subnets
Existing subnets, existing nsgs, and lb1 & lb2 with AD subnets
Existing subnets, existing nsgs, and fss
Existing subnets, existing nsgs, and fss  with atp db
Existing subnets, existing nsgs, and fss  with oci db - Testing in progress
Existing subnets, existing nsgs, and destroy

Tested negative tests
-----------------------
Apply failed due to port issues, fix and run reapply.
Apply failed due to port issues, skip the validation and rerun.

### CODE REVIEW

File: terraform/main.tf [hunk 1]
```diff
@@ -1,7 +1,31 @@
 # Copyright (c) 2023, Oracle and/or its affiliates.
 # Licensed under the Universal Permissive License v1.0 as shown at https://oss.oracle.com/licenses/upl.
 
+module "network-validation" {
```
Comment (roberto-sanchez-herrera) [inline terraform/main.tf:4]: can you please run terraform fmt in this file and all affected files in this PR?

File: terraform/modules/network-validator/variables.tf [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+variable "lb_subnet_1_id" {
+  type        = string
+  description = "The OCID of a regional or AD-specific subnet for primary load balancer"
+}
+
+variable "lb_subnet_2_id" {
+  type        = string
+  description = "The OCID of a regional or AD-specific subnet for secondary load balancer"
```
Comment (roberto-sanchez-herrera) [inline terraform/modules/network-validator/variables.tf:26]: Strictly speaking, this variable should never take the ocid of a regional subnet. It is used only for AD-specific subnets

File: terraform/modules/network-validator/variables.tf [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+variable "db_vcn_lpg_id" {
+  type        = string
+  description = "The OCID of the Local Peering Gateway (LPG) in the DB VCN to which the LPG in the WebLogic VCN will be peered. Required for VCN peering"
+}
+
+variable "wls_extern_admin_port" {
+  type        = number
+  description = "Weblogic console port"
```
Comment (roberto-sanchez-herrera) [inline terraform/modules/network-validator/variables.tf:56]: Can you change to WebLogic?

File: terraform/schema.yaml [hunk 4]
```diff
@@ -699,16 +699,16 @@ variables:
...
-    type: string
+  skip_network_validation:
+    type: boolean
+    default: false
     required: true
-    title: "Acknowledge if Existing Network is Validated"
-    description: "Enter YES to confirm if existing Virtual Cloud Network you selected is validated using the network validation script. See <a target=\"_blank\" href=\"https://docs.oracle.com/pls/topic/lookup?ctx=en/cloud/paas/weblogic-cloud/user&id=oci_network_validate\">Validate Existing Network Setup</a>"
-    pattern: "^[yY][eE][sS]$"
+    title: "Skip Network Validation"
+    description: "Network validation script to run for existing subnets"
```
Comment (roberto-sanchez-herrera) [inline terraform/schema.yaml:707]: I think the description can be improved. This description looks like this field is used to specify the script.

File: terraform/schema_14110.yaml [hunk 5]
```diff
@@ -708,16 +708,16 @@ variables:
...
-    type: string
+  skip_network_validation:
+    type: boolean
+    default: false
     required: true
-    title: "Acknowledge if Existing Network is Validated"
-    description: "Enter YES to confirm if existing Virtual Cloud Network you selected is validated using the network validation script. See <a target=\"_blank\" href=\"https://docs.oracle.com/pls/topic/lookup?ctx=en/cloud/paas/weblogic-cloud/user&id=oci_network_validate\">Validate Existing Network Setup</a>"
-    pattern: "^[yY][eE][sS]$"
+    title: "Skip Network Validation"
+    description: "Network validation script to run for existing subnets"
```
Comment (roberto-sanchez-herrera) [inline terraform/schema_14110.yaml:716]: I think the description can be improved. This description looks like this field is used to specify the script.

File: terraform/weblogic_variables.tf [hunk 6]
```diff
@@ -195,4 +195,4 @@ variable "deploy_sample_app" {
   type        = bool
   description = "Set to true to install a sample application in the WebLogic domain"
   default     = true
-}
\ No newline at end of file
+}
```
Comment (roberto-sanchez-herrera) [inline terraform/weblogic_variables.tf:198]: looks like nothing changed in this file, except a white space. If possible, I suggest to remove this change.

File: terraform/modules/network-validator/scripts/network_validation.sh [hunk 7]
```diff
@@ -0,0 +1,1103 @@
...
+  fi
+fi
+
+### Validation - Only when WLS Subnet OCID, Admin Server NSG & Managed Server NSG are provided ###
+
+if [[ -n ${WLS_SUBNET_OCID} && -n ${ADMIN_SRV_NSG_OCID} && -n ${MANAGED_SRV_NSG_OCID} ]]
+then
+  wls_subnet_cidr_block=$(oci network subnet get --subnet-id ${WLS_SUBNET_OCID} | jq -r '.data["cidr-block"]')
+
+  # Check if SSH port is open for access by WLS subnet CIDR in Admin Server NSG
```
Comment (roberto-sanchez-herrera) [inline terraform/modules/network-validator/scripts/network_validation.sh:765]: this comment and the error message says admin server nsg but the managed one is used

File: terraform/modules/network-validator/scripts/network_validation.sh [hunk 8]
```diff
@@ -0,0 +1,1103 @@
...
+  res=$(check_tcp_port_open_in_seclist_or_nsg $MANAGED_SRV_NSG_OCID "${SSH_PORT}" "$wls_subnet_cidr_block" "nsg")
+  if [[ $res -ne 0 ]]
+  then
+    echo "ERROR: Port ${SSH_PORT} is not open for access by WLS Subnet CIDR [$wls_subnet_cidr_block] in Admin Server NSG [$ADMIN_SRV_NSG_OCID]"
+    validation_return_code=2
+  fi
+
+  # Check if T3 Port is open for access by WLS subnet CIDR in Admin Server NSG
```
Comment (roberto-sanchez-herrera) [inline terraform/modules/network-validator/scripts/network_validation.sh:773]: The comment says Admin

**Review by roberto-sanchez-herrera — CHANGES_REQUESTED** [review]

File: terraform/main.tf [hunk 9]
```diff
@@ -1,7 +1,31 @@
...
+  wls_subnet_id = var.wls_subnet_id
+  bastion_subnet_id = var.is_bastion_instance_required ? var.bastion_subnet_id : ""
+  bastion_ip        = var.is_bastion_instance_required && var.existing_bastion_instance_id != "" ? data.oci_core_instance.existing_bastion_instance[0].private_ip : ""
+  lb_subnet_1_id    = var.add_load_balancer ? var.lb_subnet_1_id : ""
+  lb_subnet_2_id    = var.add_load_balancer && !local.use_regional_subnet ? var.lb_subnet_2_id : ""
+  mount_target_subnet_id    = var.add_fss ? var.mount_target_subnet_id : ""
+  atp_db_id =  !local.is_oci_db ? var.atp_db_id : ""
+  oci_db_dbsystem_id = local.is_oci_db ? var.oci_db_dbsystem_id : ""
+  oci_db_port = local.is_oci_db ? var.oci_db_port : 0
+  db_vcn_lpg_id = local.is_oci_db ? var.db_vcn_lpg_id : ""
```
Comment (roberto-sanchez-herrera) [inline terraform/main.tf:16]: this module will run when using existing subnets. But db vcn lpg id is needed only when using vcn peering, which is only used when existing subnets are not used. So, strictly speaking, this variable will always be empty.  So. I am not sure if it is worth having it

**Review by roberto-sanchez-herrera — CHANGES_REQUESTED** [review]

File: terraform/main.tf (hunk 1 above)
Comment (skommala) [reply terraform/main.tf:4]: Fixed it.

File: terraform/modules/network-validator/variables.tf (hunk 2 above)
Comment (skommala) [reply terraform/modules/network-validator/variables.tf:26]: Fixed it.

File: terraform/modules/network-validator/variables.tf (hunk 3 above)
Comment (skommala) [reply terraform/modules/network-validator/variables.tf:56]: Fixed it.

File: terraform/main.tf (hunk 9 above)
Comment (skommala) [reply terraform/main.tf:16]: Fixed it.

File: terraform/schema_14110.yaml (hunk 5 above)
Comment (skommala) [reply terraform/schema_14110.yaml:716]: changed it to "Skip running network validation script for existing subnets". I will also check with Vanaja.

File: terraform/schema.yaml (hunk 4 above)
Comment (skommala) [reply terraform/schema.yaml:707]: changed it to "Skip running network validation script for existing subnets". I will also check with Vanaja.

File: terraform/modules/network-validator/scripts/network_validation.sh (hunk 7 above)
Comment (skommala) [reply terraform/modules/network-validator/scripts/network_validation.sh:765]: Fixed.

File: terraform/modules/network-validator/scripts/network_validation.sh (hunk 8 above)
Comment (skommala) [reply terraform/modules/network-validator/scripts/network_validation.sh:773]: Fixed.

File: terraform/weblogic_variables.tf (hunk 6 above)
Comment (skommala) [reply terraform/weblogic_variables.tf:198]: ok

File: terraform/modules/network-validator/scripts/network_validation.sh [hunk 10]
```diff
@@ -512,11 +514,14 @@ This script is used to validate existing subnets, and optionally network securit
...
   -d, --ocidbid       OCI Database System OCID
+  -P, --ocidbport     OCI Database Port
   -t, --atpdbid       ATP Database OCID
   -g, --lpg           OCID of the Local Peering Gateway (LPG) in the DB VCN
   -b, --bastionsubnet Bastion Subnet OCID
-  -i, --bastionip     Bastion Host IP. Provide this if using existing bastion or new bastion with reserved IP
-  -l, --lbsubnet      Load Balancer Subnet OCID
+  -i, --bastionip     Bastion Host IP. Provide this if using existing bastion
+  -u, --lbsubnet1    Load Balancer Subnet 1 OCID
+  -v, --lbsubnet2    Load Balancer Subnet 2 OCID
```
Comment (Mahuwa-Barman) [inline terraform/modules/network-validator/scripts/network_validation.sh:523]: Please fix the indentation.

Comment (Mahuwa-Barman) [inline terraform/modules/network-validator/scripts/network_validation.sh:523]: We can mention that "Load Balancer Subnet 2 OCID" is required only for AD subnet.

File: terraform/modules/network-validator/scripts/network_validation.sh [hunk 11]
```diff
@@ -783,14 +801,15 @@ then
     # Check if DB port is open for access by WLS subnet CIDR in DB subnet/NSG
     res=$(validate_ocidb_port_access ${ocidb_subnet_ocid} ${OCIDB_OCID} ${wls_subnet_cidr_block})
     if [[ $res -ne 0 ]]; then
-      echo "ERROR: DB port ${DB_PORT} is not open for access by WLS Subnet CIDR [$wls_subnet_cidr_block] in DB Subnet [$ocidb_subnet_ocid] or in DB NSG [$ocidb_nsg_ocid]"
+      echo "ERROR: DB port ${DB_PORT} is not open for access by WLS Subnet CIDR [$wls_subnet_cidr_block] in DB Subnet [$ocidb_subnet_ocid] or in DB NSG [$ocidb_nsg_ocid]i. ${NETWORK_VALIDATION_MSG}"
```
Comment (Mahuwa-Barman) [inline terraform/modules/network-validator/scripts/network_validation.sh:804]: [$ocidb_nsg_ocid]i ?

File: terraform/modules/network-validator/scripts/network_validation.sh (hunk 10 above)
Comment (skommala) [reply terraform/modules/network-validator/scripts/network_validation.sh:523]: Fixed

Comment (skommala) [reply terraform/modules/network-validator/scripts/network_validation.sh:523]: Fixed

File: terraform/modules/network-validator/scripts/network_validation.sh (hunk 11 above)
Comment (skommala) [reply terraform/modules/network-validator/scripts/network_validation.sh:804]: Fixed typo

**Comment by Mahuwa-Barman** [comment]
Looks good to me

**Review by Mahuwa-Barman — APPROVED** [review]
Looks good to me.

**Pull request merged**